File: frontend/seer/src/socket/messages.ts
import { DecimalSchema } from "@/lib/definitions";
import z from "zod";

export const WsMessageSchema = z.object({
    type: z.string().max(100),
    payload: z.any().optional(),
});

export type WsMessage = z.infer<typeof WsMessageSchema>;

export const BalanceUpdateSchema = z.object({
    currency: z.string(),
    balance: DecimalSchema,
    version: z.number().int(),
});
export type BalanceUpdate = z.infer<typeof BalanceUpdateSchema>;

export const OutcomeUpdateSchema = z.object({
    id: z.number().int(),
    quantity: DecimalSchema,
});
export type OutcomeUpdate = z.infer<typeof OutcomeUpdateSchema>;

export const MarketUpdateSchema = z.object({
    marketID: z.uuid(),
    marketVersion: z.number().int(),
    outcomes: z.array(OutcomeUpdateSchema),
});
export type MarketUpdate = z.infer<typeof MarketUpdateSchema>;

export const WsUserSchema = z.object({
    id: z.uuid(),
    username: z.string(),
    profileImageKey: z.string().optional(),
});

export const ChatMessageSchema = z.object({
    id: z.uuid(),
    chatSlug: z.string(),
    content: z.string(),
    createdAt: z.coerce.date(),
    type: z.enum(['user', 'system']),
    user: WsUserSchema,
});


export const BetUpdateSchema = z.object({
    id: z.uuid(),
    marketID: z.uuid(),
    marketName: z.string(),
    outcomeId: z.number(),
    outcomeName: z.string(),
    wager: DecimalSchema,
    payout: DecimalSchema,
    avgPrice: DecimalSchema,
    placedAt: z.coerce.date(),
    user: WsUserSchema.optional(),
});

export const OnlineUpdateSchema = z.object({
    usersOnlineCount: z.number().int(),
});
export type OnlineUpdate = z.infer<typeof OnlineUpdateSchema>;

export const WSErrorSchema = z.object({
    error: z.string(),
});
export type WSError = z.infer<typeof WSErrorSchema>;

export const ChatRoomPrefix = 'chat:';
export const MarketRoomPrefix = 'market:';

export const getMarketRoom = (marketId: string) => `${MarketRoomPrefix}${marketId}`;
export const getChatRoom = (chatSlug: string) => `${ChatRoomPrefix}${chatSlug}`;

File: backend/internal/ws/messages.go
package ws

import (
	"seer/internal/numeric"
	"time"

	"github.com/google/uuid"
)

type UserState struct {
	ID              uuid.UUID `json:"id"`
	Username        string    `json:"username"`
	ProfileImageKey string    `json:"profileImageKey,omitempty"`
}

type BalanceUpdate struct {
	UserID   uuid.UUID           `json:"-"`
	Currency string              `json:"currency"`
	Balance  *numeric.BigDecimal `json:"balance"`
	Version  int64               `json:"version"`
}

type OutcomeUpdate struct {
	ID       int64               `json:"id"`
	Quantity *numeric.BigDecimal `json:"quantity"`
}

type MarketUpdate struct {
	ID       uuid.UUID       `json:"marketID"`
	Version  int64           `json:"marketVersion"`
	Outcomes []OutcomeUpdate `json:"outcomes"`
}

type ChatMessage struct {
	ID        uuid.UUID `json:"id"`
	ChatSlug  string    `json:"chatSlug"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"createdAt"`
	Type      string    `json:"type"` // "user" or "system"
	User      UserState `json:"user"`
}

type OnlineUpdate struct {
	UsersOnlineCount int64 `json:"usersOnlineCount"`
}

type BetUpdate struct {
	ID          uuid.UUID           `json:"id"`
	MarketID    uuid.UUID           `json:"marketID"`
	MarketName  string              `json:"marketName"`
	MarketSlug  string              `json:"marketSlug"`
	OutcomeID   int64               `json:"outcomeId"`
	OutcomeName string              `json:"outcomeName"`
	User        *UserState          `json:"user,omitempty"`
	Wager       *numeric.BigDecimal `json:"wager"`
	Payout      *numeric.BigDecimal `json:"payout"`
	AvgPrice    *numeric.BigDecimal `json:"avgPrice"`
	PlacedAt    time.Time           `json:"placedAt"`
}

type WSError struct {
	Error string `json:"error"`
}

const (
	ChatRoomPrefix    = "chat:"
	MarketsUpdateRoom = "markets_update"
	BetsLatestRoom    = "bets:latest"
	BetsHighRoom      = "bets:high"
	BalanceRoom       = "balance"
	OnlineRoom        = "online"
)
